src/linker/Linker/SerializationMarker.cs

```diff
@@ -0,0 +1,202 @@
+			}
+		}
+
+		HashSet<TypeDefinition> _recursiveTypes;
+		HashSet<TypeDefinition> RecursiveTypes {
+			get {
+				if (_recursiveTypes == null)
+					_recursiveTypes = new HashSet<TypeDefinition> ();
+
+				return _recursiveTypes;
+			}
+		}
+
+		public SerializationMarker (LinkContext context)
+		{
+			_context = context;
+		}
+
+		public bool IsActive (SerializerKind serializerKind) => ActiveSerializers.HasFlag (serializerKind);
+
+		static DependencyKind ToDependencyKind (SerializerKind serializerKind) => serializerKind switch {
+			SerializerKind.DataContractSerializer => DependencyKind.DataContractSerialized,
+			SerializerKind.XmlSerializer => DependencyKind.XmlSerialized,
+			_ => throw new ArgumentException (nameof (SerializerKind))
+		};
+
+		public void TrackForSerialization (TypeDefinition type, ICustomAttributeProvider provider, SerializerKind serializerKind)
+		{
+			if (ActiveSerializers.HasFlag (serializerKind)) {
+				MarkRecursiveMembers (type, new DependencyInfo (ToDependencyKind (serializerKind), provider));
+				return;
+			}
+
+			if (!TrackedTypes.TryGetValue (serializerKind, out var types)) {
+				types = new HashSet<(TypeDefinition, ICustomAttributeProvider)> ();
+				TrackedTypes.Add (serializerKind, types);
+			}
+
+			types.Add ((type, provider));
+		}
+
+		public void Activate (SerializerKind serializerKind)
+		{
+			if (!Enum.IsDefined<SerializerKind> (serializerKind) || serializerKind == SerializerKind.None)
+				throw new ArgumentException ($"Unexpected serializer kind {nameof (serializerKind)}");
+
+			if (ActiveSerializers.HasFlag (serializerKind))
+				return;
+
+			ActiveSerializers |= serializerKind;
+
+			if (!TrackedTypes.TryGetValue (serializerKind, out var types))
+				return;
+
+			foreach (var (type, provider) in types)
+				MarkRecursiveMembers (type, new DependencyInfo (ToDependencyKind (serializerKind), provider));
+
+			TrackedTypes.Remove (serializerKind);
+		}
+
+		public void MarkRecursiveMembers (TypeReference typeRef, in DependencyInfo reason)
+		{
+			MarkRecursiveMembersInternal (typeRef, reason);
+		}
+
+		void MarkRecursiveMembersInternal (TypeReference typeRef, in DependencyInfo reason)
+		{
+			if (typeRef == null)
+				return;
+
+			DependencyInfo typeReason = reason;
+			while (typeRef is GenericInstanceType git) {
+				if (git.HasGenericArguments) {
+					foreach (var argType in git.GenericArguments)
+						MarkRecursiveMembersInternal (argType, new DependencyInfo (DependencyKind.GenericArgumentType, typeRef));
+				}
+				_context.Tracer.AddDirectDependency (typeRef, typeReason, marked: false);
+				typeReason = new DependencyInfo (DependencyKind.ElementType, typeRef);
+				typeRef = (typeRef as TypeSpecification).ElementType;
+			}
+			// This doesn't handle other TypeSpecs. We are only matching what xamarin-android used to do.
+
+			TypeDefinition type = typeRef.Resolve ();
+			if (type == null)
+				return;
+
+			_context.Annotations.Mark (type, typeReason);
```

Comment:
Should we move this below the hashset check? I know it will work either way, but since we already have one hashset check, we might as well avoid a second one.

Reply:
I'd prefer it above, so that we get trace output if multiple serializers cause the same type to be discovered as a root.